app/src/main/java/co/kaush/msusf/usf/UsfVmImpl.kt

```diff
@@ -1,74 +1,97 @@
 package co.kaush.msusf.usf
 
-import io.reactivex.Observable
-import io.reactivex.ObservableTransformer
-import io.reactivex.disposables.CompositeDisposable
-import io.reactivex.subjects.PublishSubject
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
 import timber.log.Timber
 
-class UsfVmImpl<E: Any, R: Any, VS: Any, VE: Any>(
-    eventToResultTransformer: ObservableTransformer<E, R>,
-    resultToViewStateTransformer: ObservableTransformer<R, VS>,
-    resultToViewEffectTransformer: ObservableTransformer<R, VE>,
-    logger: UsfVmLogger = object : UsfVmLogger {
-        override fun debug(message: String) = Timber.d(message)
-        override fun warning(message: String) = Timber.w(message)
-        override fun error(error: Throwable, message: String)  = Timber.e(error, message)
-    }
+@OptIn(ExperimentalCoroutinesApi::class)
+abstract class UsfVmImpl<E : Any, R : Any, VS : Any, VE : Any>(
+  private val viewModelScope: CoroutineScope,
+  private val processingDispatcher: CoroutineDispatcher = Dispatchers.IO,
+  logger: UsfVmLogger = object : UsfVmLogger {
+    override fun debug(message: String) = Timber.d(message)
+    override fun warning(message: String) = Timber.w(message)
+    override fun error(error: Throwable, message: String) = Timber.e(error, message)
+  }
 ) : UsfVm<E, R, VS, VE> {
 
-    override val eventSink: PublishSubject<E> = PublishSubject.create()
-    override val disposables: CompositeDisposable = CompositeDisposable()
+  // remember that a single E(vent) can yield multiple R(esult)s
+  //  so if you wish to show an initial loading state before the final result
+  //  emit multiple R(esult)s for the single E(vent)
+  abstract fun eventToResultFlow(): (E) -> Flow<R>
 
-    private val viewStateSink: Observable<VS>
-    private val viewEffectSink: Observable<VE>
+  // curiously, we don't return a Flow<VS> here
+  //  as every (R)esult will only ever be transformed into a single (V)iew(S)tate
+  //  if you want multiple (V)iew(S)tates, then emit multiple (R)esults
+  //  and transform each (R)esult accordingly to the respective (V)iew(S)tate
+  abstract fun resultToViewState(): (currentViewState: VS, result: R) -> VS
 
-    init {
-        logger.debug("------ init ${Thread.currentThread().name}")
+  // a single (R)esult can result in multiple (V)iew(E)ffects
+  //  for e.g. emit one VE for navigation and another for making an analytics call
+  //  hence a return type of Flow<VE>
+  abstract fun resultToViewEffectFlow(): (R) -> Flow<VE>
 
-        eventSink
-            .compose(eventToResultTransformer)
-            .doOnNext { logger.debug("----- result ${Thread.currentThread().name} $it") }
+  abstract val initialViewState: VS
```

Comment:
i'm still thinking about this one. i personally don't think we ever _need_ an initial view state for the pattern (as that can be captured in an "on screen load" event). 

i added this mostly cause i _need_ it for the `scan` operator below. Keeping it for now, cause i also like the idea of the implementor explicitly telling us what `VS` we should start in.